### PR TITLE
feat: add matrix-auth integration

### PR DESCRIPTION
**Pull request by amandahla**
Applicable spec: <link>

### Overview

Add matrix-auth integration that allows integrate Maubot with Synapse.

### Rationale

The first step to install a bot is create a client.

A client requires that the homeserver is configured in the Maubot configuration file.

Once is there, the user will run an action "register-client-account" that will create a user and return an access token and a device id.

See here for the entire flow:
https://docs.mau.fi/maubot/usage/basic.html#creating-clients

Extra:
- If the integrations PostgreSQL and/or matrix-auth relation data are not found, the charm is blocked and if the service is running, is stopped. Thanks @javierdelapuente for spotting this
- Integration test will be added once Synapse provides the matrix-auth endpoint.

### Juju Events Changes
_on_matrix_auth_request_processed

### Library Changes

Matrix auth library.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

No CH available yet.

### CODE REVIEW

File: lib/charms/synapse/v0/matrix_auth.py
Comment (cbartz) [inline lib/charms/synapse/v0/matrix_auth.py:1]: Is this the lib from https://github.com/canonical/synapse-operator/pull/509 ?

File: tests/unit/test_charm.py [hunk 1]
```diff
@@ -32,6 +33,28 @@ def set_postgresql_integration(harness) -> None:
...
+
+    Args:
+        harness: harness instance.
+        monkeypatch: monkeypatch instance.
+    """
+    monkeypatch.setattr(
+        MatrixAuthProviderData, "get_shared_secret", lambda *args: "test-shared-secret"
+    )
+    relation_data = {"homeserver": "https://example.com", "shared_secret_id": "test-secret-id"}
+    matrix_relation_id = harness.add_relation(  # pylint: disable=attribute-defined-outside-init
```
Comment (cbartz) [inline tests/unit/test_charm.py:47]: Why does the warning needs to be disabled here, so where is the instance attribute defined?

File: tests/unit/test_charm.py [hunk 2]
```diff
@@ -32,6 +33,28 @@ def set_postgresql_integration(harness) -> None:
     )
 
 
+def set_matrix_auth_integration(harness, monkeypatch) -> None:
```
Comment (cbartz) [inline tests/unit/test_charm.py:36]: nit: The current teams standard suggest using top-down approach for function ordering, which means that this function has to be placed below `test_matrix_credentials_registered`.

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -100,17 +125,33 @@ def test_database_created(harness):
...
     harness.update_config({"public-url": "https://example1.com"})
 
     service = harness.model.unit.get_container("maubot").get_service("maubot")
     assert service.is_running()
     assert harness.model.unit.status == ops.ActiveStatus()
+
+
+def test_matrix_credentials_registered(harness, monkeypatch):
+    """
+    arrange: initialize harness and verify that there is no credentials.
```
Comment (cbartz) [inline tests/unit/test_charm.py:147]: Where is it verified that there are no credentials?

**Review by cbartz — COMMENTED** [review]
Thank you. I have some comments/questions ;-)

File: src/charm.py [hunk 4]
```diff
@@ -83,8 +100,14 @@ def _reconcile(self) -> None:
             return
         try:
             self._configure_maubot(container)
-        except MissingPostgreSQLRelationDataError:
-            self.unit.status = ops.BlockedStatus("postgresql integration is required")
+        except MissingRelationDataError as e:
+            self.unit.status = ops.BlockedStatus(f"{e.relation_name} integration is required")
+            try:
+                container.stop(MAUBOT_NAME)
+            except RuntimeError:
```
Comment (javierdelapuente) [inline src/charm.py:107]: Are we sure that container.stop raises `RuntimeError` for some problem in the workload/service/pebble? This is a generic exception that does not look related to a service being stopped.

File: src/charm.py [hunk 5]
```diff
@@ -124,23 +151,51 @@ def _get_postgresql_credentials(self) -> str:
...
+
+        Returns:
+            matrix credentials.
+
+        Raises:
+            MissingRelationDataError: if relation is not found.
+        """
+        relation = self.model.get_relation("matrix-auth")
+        if not relation or not relation.app:
+            logging.warning("no matrix-auth relation found, getting default matrix credentials")
```
Comment (javierdelapuente) [inline src/charm.py:188]: Shouldn't this block the charm?

**Review by javierdelapuente — COMMENTED** [review]
A couple of comments/questions

File: lib/charms/synapse/v0/matrix_auth.py
Comment (amandahla) [reply lib/charms/synapse/v0/matrix_auth.py:1]: Yes, same library.

File: tests/unit/test_charm.py (hunk 1 above)
Comment (amandahla) [reply tests/unit/test_charm.py]: Good catch, the comment is not needed, removed now.

File: tests/unit/test_charm.py (hunk 2 above)
Comment (amandahla) [reply tests/unit/test_charm.py]: Thanks for spotting it, moved :)

File: tests/unit/test_charm.py (hunk 3 above)
Comment (amandahla) [reply tests/unit/test_charm.py]: I added the description but forgot to add the check :|
Fixed now.

File: src/charm.py (hunk 4 above)
Comment (amandahla) [reply src/charm.py:107]: Actually the RunTimeError was raised by harness during unit tests so I added it just in case.

Comment (amandahla) [reply src/charm.py:107]: Someone already filled an issue :)
https://github.com/canonical/operator/issues/1310

Comment (amandahla) [reply src/charm.py]: Workaround: I created a decorator for handling this exception in the unit tests.

File: lib/charms/synapse/v0/matrix_auth.py [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+
+    def set_shared_secret_id(self, model: ops.Model, relation: ops.Relation) -> None:
+        """Store the Matrix shared secret as a Juju secret.
+
+        Args:
+            model: the Juju model
+            relation: relation to grant access to the secrets to.
+        """
+        # password is always defined since pydantic guarantees it
+        password = cast(SecretStr, self.shared_secret)
```
Comment (jdkandersson) [inline lib/charms/synapse/v0/matrix_auth.py:112]: Why mark this with the `Optional` type if pydantic guarantees it is not `None`? Generally, if something can't be `None`, that should be reflected in the type

File: lib/charms/synapse/v0/matrix_auth.py [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+        # pylint doesn't like get_secret_value
+        secret_value = password.get_secret_value()  # pylint: disable=no-member
```
Comment (jdkandersson) [inline lib/charms/synapse/v0/matrix_auth.py:114]: I'm guessing the reason that pylint is not happy is because it is typed as potentially being `None`. If the type was updated to not be `None`, pydantic probably will be happy

File: lib/charms/synapse/v0/matrix_auth.py [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+        try:
+            secret = model.get_secret(label=SHARED_SECRET_LABEL)
+            secret.set_content({SHARED_SECRET_CONTENT_LABEL: secret_value})
+            # secret.id is not None at this point
+            self.shared_secret_id = cast(str, secret.id)
```
Comment (jdkandersson) [inline lib/charms/synapse/v0/matrix_auth.py:119]: Is there anything here you can think of to avoid needing to `cast`?

Comment (jdkandersson) [reply lib/charms/synapse/v0/matrix_auth.py:119]: E.g., check that it isn't None and raise a value error if it is `None`?

File: lib/charms/synapse/v0/matrix_auth.py [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+        except ops.SecretNotFoundError:
+            secret = relation.app.add_secret(
+                {SHARED_SECRET_CONTENT_LABEL: secret_value}, label=SHARED_SECRET_LABEL
+            )
+            secret.grant(relation)
+            self.shared_secret_id = cast(str, secret.id)
```
Comment (jdkandersson) [inline lib/charms/synapse/v0/matrix_auth.py:125]: looks like regardless of whether an exception is raised, this code is executed. Perhaps put it outside the `try...except` block?

File: lib/charms/synapse/v0/matrix_auth.py [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+
+    @classmethod
```
Comment (jdkandersson) [inline lib/charms/synapse/v0/matrix_auth.py:127]: What is the reason for making this a `classmethod`? Generally, class methods are intended for factory like methods that produce an instance of the class. I don't see any use of the `cls` argument?

File: lib/charms/synapse/v0/matrix_auth.py [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+    def get_shared_secret(
```
Comment (jdkandersson) [inline lib/charms/synapse/v0/matrix_auth.py:128]: It seems that this actually retrieves the password rather than the secret? Should the name of the function be updated to reflect that?

File: lib/charms/synapse/v0/matrix_auth.py [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+            secret = model.get_secret(id=shared_secret_id)
+            password = secret.get_content().get(SHARED_SECRET_CONTENT_LABEL)
+            if not password:
+                return None
+            return SecretStr(password)
+        except ops.SecretNotFoundError:
+            return None
+
+    def to_relation_data(self, model: ops.Model, relation: ops.Relation) -> Dict[str, str]:
+        """Convert an instance of MatrixAuthProviderData to the relation representation.
```
Comment (jdkandersson) [inline lib/charms/synapse/v0/matrix_auth.py:152]: It seems that this method does something more than just convert, it also calls `set_shared_secret_id`. Just looking at the function name, I might not expect that to happen. Should the name of the function be updated?

File: lib/charms/synapse/v0/matrix_auth.py [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+        Args:
+            relation: the relation.
+
+        Returns:
+            A MatrixAuthProviderData instance.
+
+        Raises:
+            ValueError: if the value is not parseable.
+        """
+        app = cast(ops.Application, relation.app)
```
Comment (jdkandersson) [inline lib/charms/synapse/v0/matrix_auth.py:177]: What is the reason for needing the cast here please?

File: lib/charms/synapse/v0/matrix_auth.py [hunk 14]
```diff
@@ -0,0 +1,476 @@
...
+        relation_data = relation.data[app]
+        shared_secret_id = (
+            (relation_data["shared_secret_id"])
+            if "shared_secret_id" in relation_data
+            else None
+        )
+        shared_secret = MatrixAuthProviderData.get_shared_secret(model, shared_secret_id)
+        homeserver = relation_data.get("homeserver")
+        if shared_secret is None or homeserver is None:
```
Comment (jdkandersson) [inline lib/charms/synapse/v0/matrix_auth.py:186]: Could this be raised above if `shared_secret_id` is `None` and then `get_shared_secret` doesn't have to handle that case? Generally, exceptions should be raised as early as possible so that the traceback has the maximum content and points to the root cause. The longer the raising is delayed, the harder it is to find the reason why the exception was raised.

**Review by jdkandersson — COMMENTED** [review]
I haven't finished the review. I have left some comments that may be applicable in the rest of the code. Let's discuss these initially, potentially update the rest of the code and then we can look at the rest

File: lib/charms/synapse/v0/matrix_auth.py [hunk 15]
```diff
@@ -0,0 +1,476 @@
+# Copyright 2024 Canonical Ltd.
+# Licensed under the Apache2.0. See LICENSE file in charm source for details.
+
+"""Library to manage the plugin integrations with the Synapse charm.
+
+This library contains the Requires and Provides classes for handling the integration
+between an application and a charm providing the `matrix_plugin` integration.
```
[the remark this answers is not in the document]
Comment (amandahla) [reply lib/charms/synapse/v0/matrix_auth.py:7]: Hi, thanks for taking the time to review this library but it's not part of this repo.
It was added here as part of Synapse charm.
https://github.com/canonical/synapse-operator/pull/509

File: src/charm.py [hunk 16]
```diff
@@ -87,19 +103,32 @@ def _configure_maubot(self) -> None:
...
+                self.container.stop(MAUBOT_NAME)
+            except RuntimeError as re:
+                if str(re) == '400 Bad Request: service "maubot" does not exist':
+                    # Remove this once Harness is fixed
+                    # See https://github.com/canonical/operator/issues/1310
+                    pass
+                else:
+                    raise re
+            except (ops.pebble.ChangeError, ops.pebble.APIError) as pe:
+                logging.exception("failed to stop maubot", exc_info=pe)
```
Comment (jdkandersson) [inline src/charm.py:131]: Should the charm go into error state here since it failed to stop maubot when it needed to?

File: tests/unit/test_charm.py
Comment (jdkandersson) [inline tests/unit/test_charm.py:1]: Should there be an integration test?

File: src/charm.py [hunk 17]
```diff
@@ -32,8 +33,18 @@
 NGINX_NAME = "nginx"
 
 
-class MissingPostgreSQLRelationDataError(Exception):
-    """Custom exception to be raised in case of malformed/missing Postgresql relation data."""
+class MissingRelationDataError(Exception):
+    """Custom exception to be raised in case of malformed/missing relation data."""
+
+    def __init__(self, message: str, relation_name: str = "") -> None:
```
[the remark this answers is not in the document]
Comment (amandahla) [reply src/charm.py:39]: No reason, it was a slip, removed. Thanks for spotting it.

**Review by javierdelapuente — APPROVED** [review]

File: tests/unit/test_charm.py [hunk 18]
```diff
@@ -91,6 +74,8 @@ def test_database_created(harness):
     assert: postgresql credentials are set as expected.
     """
     harness.begin_with_initial_hooks()
+    with pytest.raises(MissingRelationDataError):
```
Comment (Thanhphan1147) [inline tests/unit/test_charm.py:77]: Maybe we can also check that the relation name set in the exception is correct?

File: src/charm.py [hunk 19]
```diff
@@ -32,8 +33,18 @@
...
+
+    def __init__(self, message: str, relation_name: str) -> None:
+        """Init custom exception.
+
+        Args:
+            message: Exception message.
+            relation_name: Relation name that raised the exception.
+        """
+        super().__init__(message)
+        self.relation_name = relation_name
```
Comment (Thanhphan1147) [inline src/charm.py:47]: I don't see this attributes used anywhere, is it necessary?

**Review by Thanhphan1147 — APPROVED** [review]
LGTM, some minor comments

File: tests/unit/test_charm.py (hunk 18 above)
Comment (amandahla) [reply tests/unit/test_charm.py:77]: Good point, I'll address that in this PR:
https://github.com/canonical/maubot-operator/pull/17

File: src/charm.py (hunk 19 above)
Comment (amandahla) [reply src/charm.py:47]: relation_name is used only while logging messages or setting the BlockedStatus message :)
